zerver/lib/markdown/__init__.py

```diff
@@ -1595,10 +1595,54 @@ def __init__(self, compiled_re: Pattern, md: markdown.Markdown) -> None:
 
 
 class AutoLink(CompiledPattern):
+    def shorten_links(self, href: str) -> Optional[str]:
+        parts = urllib.parse.urlparse(href)
+        scheme, netloc, path, params, query, fragment = parts
+        if scheme == "https" and netloc in ["github.com"]:
+            # Split the path to extract our 4 variables.
+
+            # To do it cleanly without multiple if branches based on which of these
+            # variables are present, we here add a list of ["", "", ""...]
+            # to the result of path.split, which at worst can be []. We also remove
+            # the first empty string we'd get from "/foo/bar".split("/").
+
+            # Example path: "/foo/bar" output: ["foo", "bar", "", "", ""]
+            #         path: ""         output: ["", "", "", "", ""]
+            organisation, repository, artifact, value, remaining_path = (
+                path.split("/", 5)[1:] + [""] * 5
+            )[:5]
+
+            # Decide what type of links to shorten.
+            if not organisation or not repository or not artifact or not value:
+                return None
+            repo_short_text = "{}/{}".format(organisation, repository)
+            # Some urls like https://github.com/zulip/zulip/pull/13168/commits can have
+            # some components after the artifact value as well. We should append that
+            # to the result if it is present.
+            remaining_text = ""
+            if remaining_path:
+                remaining_text += "/" + remaining_path
+            if fragment:
+                remaining_text += "#" + fragment
+
+            if netloc == "github.com":
+                return self.shorten_github_links(artifact, repo_short_text, value, remaining_text)
+        return None
+
+    def shorten_github_links(
+        self, artifact: str, repo_short_text: str, value: str, remaining_text: str
+    ) -> Optional[str]:
+        if artifact in ["pull", "issues"]:
+            return "{}#{}{}".format(repo_short_text, value, remaining_text)
+        if artifact == "commit":
+            return "{}@{}".format(repo_short_text, value[0:10])
+        return None
+
     def handleMatch(self, match: Match[str]) -> ElementStringNone:
         url = match.group("url")
         db_data = self.md.zulip_db_data
-        return url_to_a(db_data, url)
+        shortened_text = self.shorten_links(url)
+        return url_to_a(db_data, url, shortened_text)
```

Comment:
I think it would be better for this to have an `if shortened_text is not None` here, and two `url_to_a` calls, since it's a little subtle that `url_to_a` treats `None` as no parameter for `shortened_text`.